Error using reactive streams toStreamBuffered with newer versions of the AWS SDK

I have a method which uses the S3 transfer manager to return a publisher:
```scala
  def download(bucket: String, key: String): IO[Publisher[ByteBuffer]] = {
    val getObjectRequest = GetObjectRequest.builder.key(key).bucket(bucket).build
    val downloadRequest = DownloadRequest.builder
      .getObjectRequest(getObjectRequest)
      .responseTransformer(AsyncResponseTransformer.toPublisher[GetObjectResponse])
      .build()
    val completableFuture = transferManager
      .download(downloadRequest)
      .completionFuture()
    IO.fromCompletableFuture(IO(completableFuture)).map(_.result())
  }
```

I'm then calling this and calling `toStreamBuffered`
```scala
s3.download(downloadBucket, downloadBucketKey)
      .flatMap(
        _.toStreamBuffered[IO](10 * 1024)
```

With recent (I'm not 100% sure which one) versions of the AWS SDK, I'm now getting the following error:
```
Exception in thread "Thread-71" java.lang.NullPointerException: Cannot invoke "org.reactivestreams.Subscription.request(long)" because "this.subscription" is null
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.maybeRequestMore(ByteBufferStoringSubscriber.java:205)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.addBufferedDataAmount(ByteBufferStoringSubscriber.java:200)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.transferTo(ByteBufferStoringSubscriber.java:102)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtRequestBodyStreamAdapter.sendRequestBody(S3CrtRequestBodyStreamAdapter.java:47)

```

SDK version 2.20.69 is working but 2.21.42 is not. The only recent change in the SDK I can see is https://github.com/aws/aws-sdk-java-v2/pull/4308
If I revert this change locally and run my code again, it's working so it looks like that change to the AWS SDK is at least partly responsible. 

Do you know what is causing this error or if there are any workarounds I can use?

Thanks

If I create the publisher like that and then use the `FlowAdapters` to convert from the `Flow.Publisher` to `reactivestreams.Publisher` I'm getting a `StackOverflowError` 

```
Exception in thread "Thread-1" java.lang.StackOverflowError
	at software.amazon.awssdk.transfer.s3.internal.progress.DefaultTransferProgress.lambda$updateAndGet$0(DefaultTransferProgress.java:51)
	at java.base/java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:210)
	at software.amazon.awssdk.transfer.s3.internal.progress.DefaultTransferProgress.updateAndGet(DefaultTransferProgress.java:51)
	at software.amazon.awssdk.transfer.s3.internal.progress.TransferProgressUpdater.incrementBytesTransferred(TransferProgressUpdater.java:183)
	at software.amazon.awssdk.transfer.s3.internal.progress.TransferProgressUpdater.access$100(TransferProgressUpdater.java:42)
	at software.amazon.awssdk.transfer.s3.internal.progress.TransferProgressUpdater$1.subscriberOnNext(TransferProgressUpdater.java:88)
	at software.amazon.awssdk.transfer.s3.internal.progress.TransferProgressUpdater$1.subscriberOnNext(TransferProgressUpdater.java:78)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber.lambda$onNext$0(SubscriberListener.java:84)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber.invoke(SubscriberListener.java:102)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber.onNext(SubscriberListener.java:84)
	at uk.gov.nationalarchives.LambdaRunner$$anon$1$$anon$2.request(LambdaRunner.scala:50)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber$NotifyingSubscription.request(SubscriberListener.java:118)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.maybeRequestMore(ByteBufferStoringSubscriber.java:205)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.addBufferedDataAmount(ByteBufferStoringSubscriber.java:200)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.onNext(ByteBufferStoringSubscriber.java:182)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.onNext(ByteBufferStoringSubscriber.java:36)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber.onNext(SubscriberListener.java:85)
	at uk.gov.nationalarchives.LambdaRunner$$anon$1$$anon$2.request(LambdaRunner.scala:50)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber$NotifyingSubscription.request(SubscriberListener.java:118)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.maybeRequestMore(ByteBufferStoringSubscriber.java:205)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.addBufferedDataAmount(ByteBufferStoringSubscriber.java:200)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.onNext(ByteBufferStoringSubscriber.java:182)
	at software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber.onNext(ByteBufferStoringSubscriber.java:36)
	at software.amazon.awssdk.core.async.listener.SubscriberListener$NotifyingSubscriber.onNext(SubscriberListener.java:85)
	at uk.gov.nationalarchives.LambdaRunner$$anon$1$$anon$2.request(LambdaRunner.scala:50)

	...repeats a lot

	Caused by: software.amazon.awssdk.core.exception.SdkClientException: Failed to send the request: A callback has reported failure.
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:47)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.handleError(S3CrtResponseHandlerAdapter.java:139)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onFinished(S3CrtResponseHandlerAdapter.java:99)
	at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onFinished(S3MetaRequestResponseHandlerNativeAdapter.java:24)
```

I don't know if this helps but it does work using this https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/utils/async/SimplePublisher.html

```scala
  val publisher = new SimplePublisher[ByteBuffer]
  publisher.send(ByteBuffer.wrap("test".getBytes))
  val requestBody = AsyncRequestBody.fromPublisher(publisher)
  transferManager.upload(UploadRequest.builder.requestBody(requestBody).build()).completionFuture().get
  publisher.complete()
```

--- a/core/shared/src/main/scala/fs2/interop/flow/StreamPublisher.scala
+++ b/core/shared/src/main/scala/fs2/interop/flow/StreamPublisher.scala
@@ -47,18 +47,18 @@ private[flow] sealed abstract class StreamPublisher[F[_], A] private (
 )(implicit
     F: Async[F]
 ) extends Publisher[A] {
-  protected def runSubscription(subscribe: F[Unit]): Unit
+  protected def runSubscription(run: F[Unit]): Unit
 
   override final def subscribe(subscriber: Subscriber[_ >: A]): Unit = {
     requireNonNull(
       subscriber,
       "The subscriber provided to subscribe must not be null"
     )
-    try
-      runSubscription(
-        StreamSubscription.subscribe(stream, subscriber)
-      )
-    catch {
+    try {
+      val subscription = StreamSubscription(stream, subscriber)
+      subscriber.onSubscribe(subscription)
+      runSubscription(subscription.run)
+    } catch {
       case _: IllegalStateException | _: RejectedExecutionException =>
         subscriber.onSubscribe(new Subscription {
           override def cancel(): Unit = ()
@@ -72,12 +72,12 @@ private[flow] sealed abstract class StreamPublisher[F[_], A] private (
 private[flow] object StreamPublisher {
   private final class DispatcherStreamPublisher[F[_], A](
       stream: Stream[F, A],
-      startDispatcher: Dispatcher[F]
+      dispatcher: Dispatcher[F]
   )(implicit
       F: Async[F]
   ) extends StreamPublisher[F, A](stream) {
-    override protected final def runSubscription(subscribe: F[Unit]): Unit =
-      startDispatcher.unsafeRunAndForget(subscribe)
+    override protected final def runSubscription(run: F[Unit]): Unit =
+      dispatcher.unsafeRunAndForget(run)
   }
 
   private final class IORuntimeStreamPublisher[A](
@@ -85,8 +85,8 @@ private[flow] object StreamPublisher {
   )(implicit
       runtime: IORuntime
   ) extends StreamPublisher[IO, A](stream) {
-    override protected final def runSubscription(subscribe: IO[Unit]): Unit =
-      subscribe.unsafeRunAndForget()
+    override protected final def runSubscription(run: IO[Unit]): Unit =
+      run.unsafeRunAndForget()
   }
 
   def apply[F[_], A](
@@ -94,8 +94,8 @@ private[flow] object StreamPublisher {
   )(implicit
       F: Async[F]
   ): Resource[F, StreamPublisher[F, A]] =
-    Dispatcher.parallel[F](await = true).map { startDispatcher =>
-      new DispatcherStreamPublisher(stream, startDispatcher)
+    Dispatcher.parallel[F](await = true).map { dispatcher =>
+      new DispatcherStreamPublisher(stream, dispatcher)
     }
 
   def unsafe[A](

--- a/core/shared/src/main/scala/fs2/interop/flow/StreamPublisher.scala
+++ b/core/shared/src/main/scala/fs2/interop/flow/StreamPublisher.scala
@@ -29,7 +29,7 @@ import cats.effect.std.Dispatcher
 import cats.effect.unsafe.IORuntime
 
 import java.util.Objects.requireNonNull
-import java.util.concurrent.Flow.{Publisher, Subscriber, Subscription}
+import java.util.concurrent.Flow.{Publisher, Subscriber}
 import java.util.concurrent.RejectedExecutionException
 import scala.util.control.NoStackTrace
 
@@ -54,16 +54,12 @@ private[flow] sealed abstract class StreamPublisher[F[_], A] private (
       subscriber,
       "The subscriber provided to subscribe must not be null"
     )
-    try {
-      val subscription = StreamSubscription(stream, subscriber)
-      subscriber.onSubscribe(subscription)
+    val subscription = StreamSubscription(stream, subscriber)
+    subscriber.onSubscribe(subscription)
+    try
       runSubscription(subscription.run)
-    } catch {
+    catch {
       case _: IllegalStateException | _: RejectedExecutionException =>
-        subscriber.onSubscribe(new Subscription {
-          override def cancel(): Unit = ()
-          override def request(x$1: Long): Unit = ()
-        })
         subscriber.onError(StreamPublisher.CanceledStreamPublisherException)
     }
   }

--- a/core/shared/src/main/scala/fs2/interop/flow/StreamSubscription.scala
+++ b/core/shared/src/main/scala/fs2/interop/flow/StreamSubscription.scala
@@ -173,10 +173,10 @@ private[flow] final class StreamSubscription[F[_], A] private (
 private[flow] object StreamSubscription {
   private final val Sentinel = () => ()
 
-  // SIDE-EFFECTING!
-  // We need this constructor to don't be wrapped on F,
-  // Because some consumers expect we call Subscriber.onSubscribe(StreamSubscription)
-  // Before returning from StreamPublisher.subscribe(Subscriber)
+  // UNSAFE + SIDE-EFFECTING!
+  // We cannot wrap this constructor in F[_].
+  // Some consumers expect we call Subscriber.onSubscribe(StreamSubscription)
+  //   before returning from StreamPublisher.subscribe(Subscriber).
   // See https://github.com/typelevel/fs2/issues/3359
   def apply[F[_], A](
       stream: Stream[F, A],

--- a/core/shared/src/main/scala/fs2/interop/flow/StreamSubscription.scala
+++ b/core/shared/src/main/scala/fs2/interop/flow/StreamSubscription.scala
@@ -39,7 +39,7 @@ import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
   */
 private[flow] final class StreamSubscription[F[_], A] private (
     stream: Stream[F, A],
-    sub: Subscriber[A],
+    subscriber: Subscriber[A],
     requests: AtomicLong,
     resume: AtomicReference[() => Unit],
     canceled: AtomicReference[() => Unit]
@@ -50,12 +50,12 @@ private[flow] final class StreamSubscription[F[_], A] private (
   // Ensure we are on a terminal state; i.e. call `cancel`, before signaling the subscriber.
   private def onError(ex: Throwable): Unit = {
     cancel()
-    sub.onError(ex)
+    subscriber.onError(ex)
   }
 
   private def onComplete(): Unit = {
     cancel()
-    sub.onComplete()
+    subscriber.onComplete()
   }
 
   // This is a def rather than a val, because it is only used once.
@@ -100,7 +100,7 @@ private[flow] final class StreamSubscription[F[_], A] private (
       stream
         .through(subscriptionPipe)
         .chunks
-        .foreach(chunk => F.delay(chunk.foreach(sub.onNext)))
+        .foreach(chunk => F.delay(chunk.foreach(subscriber.onNext)))
         .compile
         .drain
 
@@ -173,28 +173,37 @@ private[flow] final class StreamSubscription[F[_], A] private (
 private[flow] object StreamSubscription {
   private final val Sentinel = () => ()
 
-  // Mostly for testing purposes.
-  def apply[F[_], A](stream: Stream[F, A], subscriber: Subscriber[A])(implicit
+  // SIDE-EFFECTING!
+  // We need this constructor to don't be wrapped on F,
+  // Because some consumers expect we call Subscriber.onSubscribe(StreamSubscription)
+  // Before returning from StreamPublisher.subscribe(Subscriber)
+  // See https://github.com/typelevel/fs2/issues/3359
+  def apply[F[_], A](
+      stream: Stream[F, A],
+      subscriber: Subscriber[A]
+  )(implicit
       F: Async[F]
-  ): F[StreamSubscription[F, A]] =
-    F.delay {
-      val requests = new AtomicLong(0L)
-      val resume = new AtomicReference(Sentinel)
-      val canceled = new AtomicReference(Sentinel)
+  ): StreamSubscription[F, A] = {
+    val requests = new AtomicLong(0L)
+    val resume = new AtomicReference(Sentinel)
+    val canceled = new AtomicReference(Sentinel)
 
-      new StreamSubscription(
-        stream,
-        subscriber,
-        requests,
-        resume,
-        canceled
-      )
-    }
+    new StreamSubscription(
+      stream,
+      subscriber,
+      requests,
+      resume,
+      canceled
+    )
+  }
 
-  def subscribe[F[_], A](stream: Stream[F, A], subscriber: Subscriber[A])(implicit
+  def subscribe[F[_], A](
+      stream: Stream[F, A],
+      subscriber: Subscriber[A]
+  )(implicit
       F: Async[F]
   ): F[Unit] =
-    apply(stream, subscriber).flatMap { subscription =>
+    F.delay(apply(stream, subscriber)).flatMap { subscription =>
       F.delay(subscriber.onSubscribe(subscription)) >>
         subscription.run
     }

--- a/core/shared/src/test/scala/fs2/interop/flow/CancellationSpec.scala
+++ b/core/shared/src/test/scala/fs2/interop/flow/CancellationSpec.scala
@@ -52,7 +52,7 @@ class CancellationSpec extends Fs2Suite {
     IO(new AtomicBoolean(false))
       .flatMap { flag =>
         val subscriber = new DummySubscriber(flag, program)
-        StreamSubscription(s, subscriber).flatMap { subscription =>
+        IO(StreamSubscription(s, subscriber)).flatMap { subscription =>
           (
             subscription.run,
             IO(subscriber.onSubscribe(subscription))

